wgpu-core/src/hub.rs

```diff
@@ -759,6 +792,141 @@ pub trait GfxBackend: hal::Backend {
     fn get_surface_mut(surface: &mut Surface) -> &mut Self::Surface;
 }
 
+pub trait HasLabel {
```

Comment:
It seems to me that this trait repeats logic that we pretty much have already in the code.

First, the `label` thing. We have `Borrow<RefCount>` implemented for all the resources, and all of them just return `&life_guard.ref_count`. So we should instead probably do something like:

```rust
trait Resource {
  const TYPE: &'static str;
  fn life_guard(&self) -> &LifeGuard;
}
```

We'd then use this trait in the state tracker instead of relying on `Borrow<RefCount>`.

Secondly, the "type" thing is already specified in `Hub::new` function when registries are created. We should unite the two, possibly by having something like this:
```rust
impl<T: Resource, I: TypedId, F: IdentityHandlerFactory<I>> Registry<T, I, F> {
    fn new_resource(backend: Backend, factory: &F) -> Self {
      Self::new(backend, factory, T::TYPE)
    }
}
```
And then call `Registry::new_resource` instead of `Registry::new` for everything that implements `Resource`.

Reply:
There are few of them that don't have `LifeGuard`, quickly glancing at least `CommandBuffer` and  `BindGroupLayout`.

I'm not sure if there is really a semantic relation with `LifeGuard` and having a label, or is it just something convenient many of them have and could be coerced to function as something common for all of them? Might just be a naming thing.

The type name makes sense though.

Reply:
Just convenience, I suppose. I don't feel strongly about this. If you feel that it shouldn't be a part of `LifeGuard`, feel free to discard that idea.

Reply:
Ooh, didn't notice that the `kind` str is already given at `Registry::new`, should I just ask it from the `Storage`?

Reply:
That's one way to reduce duplication, I suppose.
What I suggested above was to use the `Resource::TYPE` to *give* it to the storage, instead